File: cmd/pull.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/spf13/cobra"

	"github.com/pcr-developers/cli/internal/auth"
	"github.com/pcr-developers/cli/internal/store"
	"github.com/pcr-developers/cli/internal/supabase"
)

var pullCmd = &cobra.Command{
	Use:   "pull [remote-id]",
	Short: "Restore a pushed bundle to local drafts",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		remoteID := args[0]

		a := auth.Load()
		if a == nil {
			return fmt.Errorf("not logged in — run `pcr login`")
		}

		bundle, err := supabase.PullBundle(a.Token, remoteID)
		if err != nil {
			return fmt.Errorf("failed to fetch bundle: %w", err)
		}

		// The bundle contains an items array of PromptRecords
		itemsRaw, _ := bundle["items"]
		itemsJSON, _ := json.Marshal(itemsRaw)
		var items []supabase.PromptRecord
		if err := json.Unmarshal(itemsJSON, &items); err != nil {
			return fmt.Errorf("failed to parse bundle items: %w", err)
		}

		restored := 0
		for _, item := range items {
			if err := store.SaveDraft(item, nil); err == nil {
				restored++
			}
		}

		fmt.Fprintf(os.Stderr, "PCR: Restored %d prompt%s from bundle %s\n",
			restored, plural(restored), remoteID)
		return nil
	},
}

File: crates/pcr-core/src/commands/pull.rs
//! `pcr pull`. Mirrors `cli/cmd/pull.go`.

use crate::agent::{self, OutputMode};
use crate::auth;
use crate::display;
use crate::entry::PullArgs;
use crate::exit::ExitCode;
use crate::store;
use crate::supabase::{self, PromptRecord};
use crate::util::text::{parse_first_index, plural};

pub fn run(_mode: OutputMode, args: PullArgs) -> ExitCode {
    let Some(a) = auth::load() else {
        display::eprintln("not logged in — run `pcr login`");
        return ExitCode::AuthRequired;
    };

    let mut remote_id = args.remote_id.unwrap_or_default();

    if remote_id.is_empty() {
        let pushed = match store::list_pushed_commits() {
            Ok(v) => v,
            Err(e) => {
                display::print_error("pull", &e.to_string());
                return ExitCode::GenericError;
            }
        };
        if pushed.is_empty() {
            display::eprintln("PCR: No pushed prompt bundles found.");
            return ExitCode::Success;
        }
        display::eprintln("Pushed prompt bundles:\n");
        for (i, b) in pushed.iter().enumerate() {
            display::eprintln(&format!(
                "  [{}] {:?}  remote: {}",
                i + 1,
                b.message,
                b.remote_id
            ));
        }
        display::eprintln("");
        if !agent::is_interactive_terminal() {
            display::eprintln("PCR: Interactive mode not available in this terminal.");
            display::eprintln("     Use flags: pcr pull <bundle-id>");
            return ExitCode::InteractiveUnavailable;
        }
        display::eprint("Select bundle to pull [number]: ");
        let mut buf = String::new();
        if std::io::stdin().read_line(&mut buf).is_err() {
            display::eprintln("PCR: Nothing pulled.");
            return ExitCode::Success;
        }
        let Some(idx) = parse_first_index(buf.trim(), pushed.len()) else {
            display::eprintln("PCR: Nothing pulled.");
            return ExitCode::Success;
        };
        remote_id = pushed[idx].remote_id.clone();
    }

    if remote_id.is_empty() {
        display::print_error("pull", "no remote ID");
        return ExitCode::Usage;
    }

    let bundle = match supabase::pull_bundle(&a.token, &remote_id) {
        Ok(v) => v,
        Err(e) => {
            display::print_error("pull", &format!("failed to fetch bundle: {e}"));
            return ExitCode::Network;
        }
    };
    let items_raw = bundle
        .get("items")
        .cloned()
        .unwrap_or(serde_json::Value::Null);
    let items: Vec<PromptRecord> = match serde_json::from_value(items_raw) {
        Ok(v) => v,
        Err(e) => {
            display::print_error("pull", &format!("failed to parse bundle items: {e}"));
            return ExitCode::GenericError;
        }
    };

    let mut restored = 0usize;
    for item in items {
        if store::save_draft(&item, &[], "", "").is_ok() {
            restored += 1;
        }
    }
    display::eprintln(&format!(
        "PCR: Restored {} prompt{} from prompt bundle {}",
        restored,
        plural(restored),
        remote_id
    ));
    ExitCode::Success
}
